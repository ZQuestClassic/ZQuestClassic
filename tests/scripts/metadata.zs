/** COMMENT IGNORED */
// COMMENT IGNORED
int postfix = 1; // THIS COMMENT USED #1
int postfix2 = postfix; // THIS COMMENT USED #2
/** x marks the spot
 * still line one still line one still line one still line one still line one
 * 
 * line two
 * line two still
 *
 * line three
 *
 * - list item 1
 * - list item 2
 * - list item 3
*/
int x = 1;
// hello world
int y, z = 2;

// A car, duh
class Car {
    /** How fast it goes. See {@link vroom|the vroom method} or {@link speed} or [x] */
    int speed;

    // a ctor with 1 param
    Car(int speed) {
        this->speed = speed;
    }

	/** Go fast */
    void vroom() {
        speed = 1;
    }

    void vroom2() {
        vroom();
    }
}

class IhaveAdefCtor {
}

// how fancy something looks
enum Brand {
    // top class
    Fancy,
    // absolute trash
    Dull,
    Mid // meh
};

// a ffc script called Metadata
ffc script Metadata {
    // run me
    void run(int radius, int speed, int angle, int radius2, int angle2) {
        printf("%d %d\n", postfix + postfix2, 1 + x + 2);
        utils::fn(postfix);
        // i am a car
        // {@link lololol|i dont exist}
        auto c = new Car(2);
        printf("%d %d %d\n", c->speed, radius, utils::hmm + y + z + Dull + Fancy + Mid);
        auto d = new IhaveAdefCtor();
        Car c2 = new Car(2); // TODO: let's make `Car` in type decl point to the class defn. also do same for auto?
    }
}

// a namespace for cool things
namespace utils {
    /** A lovely function. See {@link utils::fn} or {@link x} or {@link fn} or {@link Car} */
    void fn(int a) {

    }

    int hmm; // hmmmmmm See {@link utils::fn} or {@link x} or {@link fn}
}
