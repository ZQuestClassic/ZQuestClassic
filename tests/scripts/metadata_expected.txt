Compiling 'metadata.zs'
Pass 1: Parsing
Pass 2: Preprocessing
Pass 3: Registration
Pass 4: Analyzing Code
Pass 5: Checking code paths
Pass 6: Generating object code
Pass 7: Assembling
Debug: could not resolve symbol "bitmap::Width"
Debug: could not resolve symbol "lololol"
Debug: could not resolve symbol "c->speed"
Debug: could not resolve symbol "bitmap::Width"
Success!
Compile finished with exit code '0' (success)
=== METADATA (elided)
{
  "currentFileSymbols": [
    {
      "name": "Fancy",
      "kind": 13,
      "range": {
        "start": {
          "line": 47,
          "character": 4
        },
        "end": {
          "line": 47,
          "character": 9
        }
      },
      "selectionRange": {
        "start": {
          "line": 47,
          "character": 4
        },
        "end": {
          "line": 47,
          "character": 9
        }
      },
      "children": []
    },
    {
      "name": "Dull",
      "kind": 13,
      "range": {
        "start": {
          "line": 49,
          "character": 4
        },
        "end": {
          "line": 49,
          "character": 8
        }
      },
      "selectionRange": {
        "start": {
          "line": 49,
          "character": 4
        },
        "end": {
          "line": 49,
          "character": 8
        }
      },
      "children": []
    },
    {
      "name": "Mid",
      "kind": 13,
      "range": {
        "start": {
          "line": 50,
          "character": 4
        },
        "end": {
          "line": 50,
          "character": 7
        }
      },
      "selectionRange": {
        "start": {
          "line": 50,
          "character": 4
        },
        "end": {
          "line": 50,
          "character": 7
        }
      },
      "children": []
    },
    {
      "name": "postfix",
      "kind": 13,
      "range": {
        "start": {
          "line": 2,
          "character": 4
        },
        "end": {
          "line": 2,
          "character": 15
        }
      },
      "selectionRange": {
        "start": {
          "line": 2,
          "character": 4
        },
        "end": {
          "line": 2,
          "character": 11
        }
      },
      "children": []
    },
    {
      "name": "postfix2",
      "kind": 13,
      "range": {
        "start": {
          "line": 3,
          "character": 4
        },
        "end": {
          "line": 3,
          "character": 22
        }
      },
      "selectionRange": {
        "start": {
          "line": 3,
          "character": 4
        },
        "end": {
          "line": 3,
          "character": 12
        }
      },
      "children": []
    },
    {
      "name": "x",
      "kind": 13,
      "range": {
        "start": {
          "line": 16,
          "character": 4
        },
        "end": {
          "line": 16,
          "character": 9
        }
      },
      "selectionRange": {
        "start": {
          "line": 16,
          "character": 4
        },
        "end": {
          "line": 16,
          "character": 5
        }
      },
      "children": []
    },
    {
      "name": "y",
      "kind": 13,
      "range": {
        "start": {
          "line": 18,
          "character": 4
        },
        "end": {
          "line": 18,
          "character": 5
        }
      },
      "selectionRange": {
        "start": {
          "line": 18,
          "character": 4
        },
        "end": {
          "line": 18,
          "character": 5
        }
      },
      "children": []
    },
    {
      "name": "z",
      "kind": 13,
      "range": {
        "start": {
          "line": 18,
          "character": 7
        },
        "end": {
          "line": 18,
          "character": 12
        }
      },
      "selectionRange": {
        "start": {
          "line": 18,
          "character": 7
        },
        "end": {
          "line": 18,
          "character": 8
        }
      },
      "children": []
    },
    {
      "name": "utils",
      "kind": 3,
      "range": {
        "start": {
          "line": 73,
          "character": 0
        },
        "end": {
          "line": 80,
          "character": 1
        }
      },
      "selectionRange": {
        "start": {
          "line": 73,
          "character": 10
        },
        "end": {
          "line": 73,
          "character": 15
        }
      },
      "children": [
        {
          "name": "hmm",
          "kind": 13,
          "range": {
            "start": {
              "line": 79,
              "character": 8
            },
            "end": {
              "line": 79,
              "character": 11
            }
          },
          "selectionRange": {
            "start": {
              "line": 79,
              "character": 8
            },
            "end": {
              "line": 79,
              "character": 11
            }
          },
          "children": []
        },
        {
          "name": "fn",
          "kind": 12,
          "range": {
            "start": {
              "line": 75,
              "character": 9
            },
            "end": {
              "line": 75,
              "character": 18
            }
          },
          "selectionRange": {
            "start": {
              "line": 75,
              "character": 9
            },
            "end": {
              "line": 75,
              "character": 11
            }
          },
          "children": []
        }
      ]
    },
    {
      "name": "Metadata",
      "kind": 2,
      "range": {
        "start": {
          "line": 54,
          "character": 0
        },
        "end": {
          "line": 70,
          "character": 1
        }
      },
      "selectionRange": {
        "start": {
          "line": 54,
          "character": 11
        },
        "end": {
          "line": 54,
          "character": 19
        }
      },
      "children": [
        {
          "name": "run",
          "kind": 12,
          "range": {
            "start": {
              "line": 56,
              "character": 9
            },
            "end": {
              "line": 56,
              "character": 71
            }
          },
          "selectionRange": {
            "start": {
              "line": 56,
              "character": 9
            },
            "end": {
              "line": 56,
              "character": 12
            }
          },
          "children": [
            {
              "name": "c",
              "kind": 13,
              "range": {
                "start": {
                  "line": 61,
                  "character": 13
                },
                "end": {
                  "line": 61,
                  "character": 27
                }
              },
              "selectionRange": {
                "start": {
                  "line": 61,
                  "character": 13
                },
                "end": {
                  "line": 61,
                  "character": 14
                }
              },
              "children": []
            },
            {
              "name": "d",
              "kind": 13,
              "range": {
                "start": {
                  "line": 63,
                  "character": 13
                },
                "end": {
                  "line": 63,
                  "character": 36
                }
              },
              "selectionRange": {
                "start": {
                  "line": 63,
                  "character": 13
                },
                "end": {
                  "line": 63,
                  "character": 14
                }
              },
              "children": []
            },
            {
              "name": "c2",
              "kind": 13,
              "range": {
                "start": {
                  "line": 64,
                  "character": 12
                },
                "end": {
                  "line": 64,
                  "character": 27
                }
              },
              "selectionRange": {
                "start": {
                  "line": 64,
                  "character": 12
                },
                "end": {
                  "line": 64,
                  "character": 14
                }
              },
              "children": []
            },
            {
              "name": "e",
              "kind": 13,
              "range": {
                "start": {
                  "line": 67,
                  "character": 12
                },
                "end": {
                  "line": 67,
                  "character": 24
                }
              },
              "selectionRange": {
                "start": {
                  "line": 67,
                  "character": 12
                },
                "end": {
                  "line": 67,
                  "character": 13
                }
              },
              "children": []
            }
          ]
        }
      ]
    },
    {
      "name": "Car",
      "kind": 5,
      "range": {
        "start": {
          "line": 21,
          "character": 0
        },
        "end": {
          "line": 39,
          "character": 1
        }
      },
      "selectionRange": {
        "start": {
          "line": 21,
          "character": 6
        },
        "end": {
          "line": 21,
          "character": 9
        }
      },
      "children": [
        {
          "name": "speed",
          "kind": 13,
          "range": {
            "start": {
              "line": 23,
              "character": 8
            },
            "end": {
              "line": 23,
              "character": 13
            }
          },
          "selectionRange": {
            "start": {
              "line": 23,
              "character": 8
            },
            "end": {
              "line": 23,
              "character": 13
            }
          },
          "children": []
        },
        {
          "name": "vroom",
          "kind": 12,
          "range": {
            "start": {
              "line": 32,
              "character": 9
            },
            "end": {
              "line": 32,
              "character": 16
            }
          },
          "selectionRange": {
            "start": {
              "line": 32,
              "character": 9
            },
            "end": {
              "line": 32,
              "character": 14
            }
          },
          "children": []
        },
        {
          "name": "vroom2",
          "kind": 12,
          "range": {
            "start": {
              "line": 36,
              "character": 9
            },
            "end": {
              "line": 36,
              "character": 17
            }
          },
          "selectionRange": {
            "start": {
              "line": 36,
              "character": 9
            },
            "end": {
              "line": 36,
              "character": 15
            }
          },
          "children": []
        },
        {
          "name": "Car",
          "kind": 9,
          "range": {
            "start": {
              "line": 27,
              "character": 4
            },
            "end": {
              "line": 27,
              "character": 18
            }
          },
          "selectionRange": {
            "start": {
              "line": 27,
              "character": 4
            },
            "end": {
              "line": 27,
              "character": 7
            }
          },
          "children": []
        }
      ]
    },
    {
      "name": "IhaveAdefCtor",
      "kind": 5,
      "range": {
        "start": {
          "line": 41,
          "character": 0
        },
        "end": {
          "line": 42,
          "character": 1
        }
      },
      "selectionRange": {
        "start": {
          "line": 41,
          "character": 6
        },
        "end": {
          "line": 41,
          "character": 19
        }
      },
      "children": []
    }
  ],
  "symbols": {
    "1345": {
      "loc": {
        "range": {
          "start": {
            "line": 2,
            "character": 4
          },
          "end": {
            "line": 2,
            "character": 15
          }
        },
        "uri": "URI"
      },
      "doc": "THIS COMMENT USED #1"
    },
    "709": {
      "loc": {
        "range": {
          "start": {
            "line": 694,
            "character": 14
          },
          "end": {
            "line": 694,
            "character": 35
          }
        },
        "uri": "URI"
      },
      "doc": "Prints the specified format_string to the output console, using the given\nargs to fill in parts of the string that are specially marked. Does not\ninclude a newline character after the string."
    },
    "1346": {
      "loc": {
        "range": {
          "start": {
            "line": 3,
            "character": 4
          },
          "end": {
            "line": 3,
            "character": 22
          }
        },
        "uri": "URI"
      },
      "doc": "THIS COMMENT USED #2"
    },
    "1347": {
      "loc": {
        "range": {
          "start": {
            "line": 16,
            "character": 0
          },
          "end": {
            "line": 16,
            "character": 9
          }
        },
        "uri": "URI"
      },
      "doc": "x marks the spot\nstill line one still line one still line one still line one still line one\nline two\nline two still\nline three\n- list item 1\n- list item 2\n- list item 3"
    },
    "2493": {
      "loc": {
        "range": {
          "start": {
            "line": 75,
            "character": 9
          },
          "end": {
            "line": 75,
            "character": 18
          }
        },
        "uri": "URI"
      },
      "doc": "A lovely function. See [`utils::fn()`](command:zscript.openLink?%7B%22file%22%3A%22%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A75%2C%22character%22%3A9%7D%2C%22end%22%3A%7B%22line%22%3A75%2C%22character%22%3A11%7D%7D%7D) or [`x`](command:zscript.openLink?%7B%22file%22%3A%22%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A16%2C%22character%22%3A4%7D%2C%22end%22%3A%7B%22line%22%3A16%2C%22character%22%3A5%7D%7D%7D) or [`fn()`](command:zscript.openLink?%7B%22file%22%3A%22%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A75%2C%22character%22%3A9%7D%2C%22end%22%3A%7B%22line%22%3A75%2C%22character%22%3A11%7D%7D%7D) or [`Car`](command:zscript.openLink?%7B%22file%22%3A%22%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A21%2C%22character%22%3A6%7D%2C%22end%22%3A%7B%22line%22%3A21%2C%22character%22%3A9%7D%7D%7D)"
    },
    "2497": {
      "loc": {
        "range": {
          "start": {
            "line": 27,
            "character": 4
          },
          "end": {
            "line": 27,
            "character": 18
          }
        },
        "uri": "URI"
      },
      "doc": "TODO: bitmap::Width does not work\na ctor with 1 param"
    },
    "Car::speed": {
      "loc": {
        "range": {
          "start": {
            "line": 23,
            "character": 4
          },
          "end": {
            "line": 23,
            "character": 13
          }
        },
        "uri": "URI"
      },
      "doc": "How fast it goes. See [`the vroom method()`](command:zscript.openLink?%7B%22file%22%3A%22%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A32%2C%22character%22%3A9%7D%2C%22end%22%3A%7B%22line%22%3A32%2C%22character%22%3A14%7D%7D%7D) or [`speed`](command:zscript.openLink?%7B%22file%22%3A%22%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A23%2C%22character%22%3A8%7D%2C%22end%22%3A%7B%22line%22%3A23%2C%22character%22%3A13%7D%7D%7D) or [`x`](command:zscript.openLink?%7B%22file%22%3A%22%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A16%2C%22character%22%3A4%7D%2C%22end%22%3A%7B%22line%22%3A16%2C%22character%22%3A5%7D%7D%7D) or [`Waitframe()`](command:zscript.openLink?%7B%22file%22%3A%22global.zh%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A22%2C%22character%22%3A14%7D%2C%22end%22%3A%7B%22line%22%3A22%2C%22character%22%3A23%7D%7D%7D) or [`bitmap`](command:zscript.openLink?%7B%22file%22%3A%22bitmap.zh%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A1%2C%22character%22%3A6%7D%2C%22end%22%3A%7B%22line%22%3A1%2C%22character%22%3A12%7D%7D%7D) or `bitmap::Width`"
    },
    "1563": {
      "loc": {
        "range": {
          "start": {
            "line": 61,
            "character": 8
          },
          "end": {
            "line": 61,
            "character": 27
          }
        },
        "uri": "URI"
      },
      "doc": "i am a car\n`i dont exist`"
    },
    "1556": {
      "loc": {
        "range": {
          "start": {
            "line": 56,
            "character": 13
          },
          "end": {
            "line": 56,
            "character": 23
          }
        },
        "uri": "URI"
      }
    },
    "1350": {
      "loc": {
        "range": {
          "start": {
            "line": 79,
            "character": 8
          },
          "end": {
            "line": 79,
            "character": 11
          }
        },
        "uri": "URI"
      },
      "doc": "hmmmmmm See [`utils::fn()`](command:zscript.openLink?%7B%22file%22%3A%22%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A75%2C%22character%22%3A9%7D%2C%22end%22%3A%7B%22line%22%3A75%2C%22character%22%3A11%7D%7D%7D) or [`x`](command:zscript.openLink?%7B%22file%22%3A%22%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A16%2C%22character%22%3A4%7D%2C%22end%22%3A%7B%22line%22%3A16%2C%22character%22%3A5%7D%7D%7D) or [`fn()`](command:zscript.openLink?%7B%22file%22%3A%22%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A75%2C%22character%22%3A9%7D%2C%22end%22%3A%7B%22line%22%3A75%2C%22character%22%3A11%7D%7D%7D)"
    },
    "1348": {
      "loc": {
        "range": {
          "start": {
            "line": 18,
            "character": 0
          },
          "end": {
            "line": 18,
            "character": 12
          }
        },
        "uri": "URI"
      },
      "doc": "hello world"
    },
    "1349": {
      "loc": {
        "range": {
          "start": {
            "line": 18,
            "character": 0
          },
          "end": {
            "line": 18,
            "character": 12
          }
        },
        "uri": "URI"
      },
      "doc": "hello world"
    },
    "1": {
      "loc": {
        "range": {
          "start": {
            "line": 49,
            "character": 4
          },
          "end": {
            "line": 49,
            "character": 8
          }
        },
        "uri": "URI"
      },
      "doc": "absolute trash"
    },
    "0": {
      "loc": {
        "range": {
          "start": {
            "line": 47,
            "character": 4
          },
          "end": {
            "line": 47,
            "character": 9
          }
        },
        "uri": "URI"
      },
      "doc": "top class"
    },
    "2": {
      "loc": {
        "range": {
          "start": {
            "line": 50,
            "character": 4
          },
          "end": {
            "line": 50,
            "character": 7
          }
        },
        "uri": "URI"
      },
      "doc": "meh"
    },
    "3551": {
      "loc": {
        "range": {
          "start": {
            "line": 41,
            "character": 6
          },
          "end": {
            "line": 41,
            "character": 19
          }
        },
        "uri": "URI"
      }
    },
    "1567": {
      "loc": {
        "range": {
          "start": {
            "line": 67,
            "character": 8
          },
          "end": {
            "line": 67,
            "character": 24
          }
        },
        "uri": "URI"
      },
      "doc": "TODO: support this in the scope lookup (see lookupDatum)\nequal to `c->speed`"
    },
    "1351": {
      "loc": {
        "range": {
          "start": {
            "line": 23,
            "character": 4
          },
          "end": {
            "line": 23,
            "character": 13
          }
        },
        "uri": "URI"
      },
      "doc": "How fast it goes. See [`the vroom method()`](command:zscript.openLink?%7B%22file%22%3A%22%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A32%2C%22character%22%3A9%7D%2C%22end%22%3A%7B%22line%22%3A32%2C%22character%22%3A14%7D%7D%7D) or [`speed`](command:zscript.openLink?%7B%22file%22%3A%22%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A23%2C%22character%22%3A8%7D%2C%22end%22%3A%7B%22line%22%3A23%2C%22character%22%3A13%7D%7D%7D) or [`x`](command:zscript.openLink?%7B%22file%22%3A%22%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A16%2C%22character%22%3A4%7D%2C%22end%22%3A%7B%22line%22%3A16%2C%22character%22%3A5%7D%7D%7D) or [`Waitframe()`](command:zscript.openLink?%7B%22file%22%3A%22global.zh%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A22%2C%22character%22%3A14%7D%2C%22end%22%3A%7B%22line%22%3A22%2C%22character%22%3A23%7D%7D%7D) or [`bitmap`](command:zscript.openLink?%7B%22file%22%3A%22bitmap.zh%22%2C%22position%22%3A%7B%22start%22%3A%7B%22line%22%3A1%2C%22character%22%3A6%7D%2C%22end%22%3A%7B%22line%22%3A1%2C%22character%22%3A12%7D%7D%7D) or `bitmap::Width`"
    },
    "2495": {
      "loc": {
        "range": {
          "start": {
            "line": 32,
            "character": 9
          },
          "end": {
            "line": 32,
            "character": 16
          }
        },
        "uri": "URI"
      },
      "doc": "Go fast"
    },
    "3870": {
      "loc": {
        "range": {
          "start": {
            "line": 27,
            "character": 8
          },
          "end": {
            "line": 27,
            "character": 17
          }
        },
        "uri": "URI"
      }
    }
  },
  "identifiers": [
    {
      "loc": {
        "line": 3,
        "character": 15,
        "length": 7
      },
      "symbol": "1345"
    },
    {
      "loc": {
        "line": 57,
        "character": 8,
        "length": 6
      },
      "symbol": "709"
    },
    {
      "loc": {
        "line": 57,
        "character": 26,
        "length": 7
      },
      "symbol": "1345"
    },
    {
      "loc": {
        "line": 57,
        "character": 36,
        "length": 8
      },
      "symbol": "1346"
    },
    {
      "loc": {
        "line": 57,
        "character": 50,
        "length": 1
      },
      "symbol": "1347"
    },
    {
      "loc": {
        "line": 58,
        "character": 15,
        "length": 2
      },
      "symbol": "2493"
    },
    {
      "loc": {
        "line": 58,
        "character": 18,
        "length": 7
      },
      "symbol": "1345"
    },
    {
      "loc": {
        "line": 61,
        "character": 21,
        "length": 3
      },
      "symbol": "2497"
    },
    {
      "loc": {
        "line": 62,
        "character": 8,
        "length": 6
      },
      "symbol": "709"
    },
    {
      "loc": {
        "line": 62,
        "character": 32,
        "length": 5
      },
      "symbol": "Car::speed"
    },
    {
      "loc": {
        "line": 62,
        "character": 29,
        "length": 1
      },
      "symbol": "1563"
    },
    {
      "loc": {
        "line": 62,
        "character": 39,
        "length": 6
      },
      "symbol": "1556"
    },
    {
      "loc": {
        "line": 62,
        "character": 54,
        "length": 3
      },
      "symbol": "1350"
    },
    {
      "loc": {
        "line": 62,
        "character": 60,
        "length": 1
      },
      "symbol": "1348"
    },
    {
      "loc": {
        "line": 62,
        "character": 64,
        "length": 1
      },
      "symbol": "1349"
    },
    {
      "loc": {
        "line": 62,
        "character": 68,
        "length": 4
      },
      "symbol": "1"
    },
    {
      "loc": {
        "line": 62,
        "character": 75,
        "length": 5
      },
      "symbol": "0"
    },
    {
      "loc": {
        "line": 62,
        "character": 83,
        "length": 3
      },
      "symbol": "2"
    },
    {
      "loc": {
        "line": 63,
        "character": 21,
        "length": 13
      },
      "symbol": "3551"
    },
    {
      "loc": {
        "line": 64,
        "character": 21,
        "length": 3
      },
      "symbol": "2497"
    },
    {
      "loc": {
        "line": 67,
        "character": 19,
        "length": 5
      },
      "symbol": "Car::speed"
    },
    {
      "loc": {
        "line": 67,
        "character": 16,
        "length": 1
      },
      "symbol": "1563"
    },
    {
      "loc": {
        "line": 68,
        "character": 8,
        "length": 7
      },
      "symbol": "0"
    },
    {
      "loc": {
        "line": 68,
        "character": 16,
        "length": 1
      },
      "symbol": "1567"
    },
    {
      "loc": {
        "line": 33,
        "character": 8,
        "length": 5
      },
      "symbol": "1351"
    },
    {
      "loc": {
        "line": 37,
        "character": 8,
        "length": 5
      },
      "symbol": "2495"
    },
    {
      "loc": {
        "line": 28,
        "character": 14,
        "length": 5
      },
      "symbol": "Car::speed"
    },
    {
      "loc": {
        "line": 28,
        "character": 22,
        "length": 5
      },
      "symbol": "3870"
    }
  ]
}
PUSHR D0; void run(int, int, int, int, int) Params
PUSHR D1
PUSHR D2
PUSHR D3
PUSHR D4
PUSHR REFFFC; void run(int, int, int, int, int) Body Start
PUSHARGSV 0 0.0006
SETR D4 SP2
ALLOCATEMEMV D2 7 0
STORE D2 0.0005
WRITEPODSTRING D2 "%d %d\n"
LOAD D2 0.0005; InlineFunc[void printf(char32)] Params Start
PUSHR D2
SETR D2 GD1
PUSHR D2
SETR D2 GD2
POP D3
ADDR D2 D3
PUSHVARGR D2
SETR D2 GD3
ADDV D2 1
ADDV D2 2
PUSHVARGR D2; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP D5; InlineFunc[void printf(char32)] Body End
LOAD D3 0.0005
DEALLOCATEMEMR D3
PUSHR CLASS_THISKEY
PUSHR D4
PUSHV 2; ClassFunc[Car Car::Car(int)] Params
CALLFUNC 0.0081; ClassFunc[Car Car::Car(int)] Constructor Call
POP D4
POP CLASS_THISKEY
STORE_OBJECT D2 0.0004
ALLOCATEMEMV D2 10 0
STORE D2 0.0003
WRITEPODSTRING D2 "%d %d %d\n"
LOAD D2 0.0003; InlineFunc[void printf(char32)] Params Start
PUSHR D2
LOAD D2 0.0004
ZCLASS_READ D2 0
PUSHVARGR D2
LOAD D2 0.0011
PUSHVARGR D2
SETR D2 GD6
PUSHR D2
SETR D2 GD4
POP D3
ADDR D2 D3
PUSHR D2
SETR D2 GD5
POP D3
ADDR D2 D3
ADDV D2 1
ADDV D2 2
PUSHVARGR D2; InlineFunc[void printf(char32)] Params End
PRINTFVARG; InlineFunc[void printf(char32)] Body Start
POP D5; InlineFunc[void printf(char32)] Body End
LOAD D3 0.0003
DEALLOCATEMEMR D3
SETV D2 0; ProtoFunc[IhaveAdefCtor IhaveAdefCtor::IhaveAdefCtor()] Set Destructor
ZCLASS_CONSTRUCT D2 {0}; ProtoFunc[IhaveAdefCtor IhaveAdefCtor::IhaveAdefCtor()] Default Construct
STORE_OBJECT D2 0.0002
PUSHR CLASS_THISKEY
PUSHR D4
PUSHV 2; ClassFunc[Car Car::Car(int)] Params
CALLFUNC 0.0081; ClassFunc[Car Car::Car(int)] Constructor Call
POP D4
POP CLASS_THISKEY
STORE_OBJECT D2 0.0001
LOAD D2 0.0004
ZCLASS_READ D2 0
STORE D2 0
LOAD D2 0; InlineFunc[int WrapDeg(int)] Params
WRAPDEGREES D2; InlineFunc[int WrapDeg(int)] Body
REF_REMOVE 0.0004
REF_REMOVE 0.0001
REF_REMOVE 0.0002
POPARGS D5 0.0012
QUIT; void run(int, int, int, int, int) Body End
SETV D2 0; Func[Car Car::Car(int)] Body Start
SETR CLASS_THISKEY2 CLASS_THISKEY
ZCLASS_CONSTRUCT CLASS_THISKEY {0.0001}
SETR D4 SP2
PUSHR CLASS_THISKEY2
LOAD D2 0
PUSHR D2
SETR D2 CLASS_THISKEY
SETR D3 D2
POP D2
ZCLASS_WRITE D3 0
POP D5
SETR D2 CLASS_THISKEY
POP CLASS_THISKEY
RETURNFUNC; Func[Car Car::Car(int)] Body End


SETV GD1 1; void run() Body Start
SETV GD2 1
SETV GD3 1
SETV GD4 0
SETV GD5 2
SETV GD6 0
QUIT; void run() Body End